### PR TITLE
Add single object support

### LINKED ISSUE #130: Avro java serialized deserialization not working
I'm working on a project that need to have a java service and intended to use avro for service to service communication as follow:

                      JAVA APP----avro Req (byte serialized)----> GO APP
                      <-----avro Rsp (byte serialized)----

I had trouble deserializing Java serialized avro and vice versa deserializing go serialized response. After investigation what I see is the following:

- Java avro serializer when serializing an avro message add a fixed header (2 bytes) (0xC301) to annouce the avro version at the beginning of the binary message
- Java avro serializer add a long (8 bytes) (ex: 0C947F601DE7CE84)  which is a magic number to announce the schema version ( it's calculated based on the schema string)

If I remove those both headers the message is working correctly and the serialization/deserialization is done correctly on both sides.

I'm not an avro expert but this seem to be part of the avro specification and I'm wondering if this is something that would be added and that you will agree to merge. If so, I can work on it with a bit of help.

I would like also to mention that it also not supported by goavro

### PR DESCRIPTION
**Pull request by dhiaayachi**
I added the support for single object through the use of reader/writer interface implementation. 
This resolve #130 
Please review, some code clean up is probably needed and also some documentation.

### CODE REVIEW

File: generator/flat/templates/record.go [hunk 1]
```diff
@@ -6,9 +6,17 @@ import (
 	"github.com/actgardner/gogen-avro/vm/types"
 	"github.com/actgardner/gogen-avro/vm"
 	"github.com/actgardner/gogen-avro/compiler"
+	"github.com/actgardner/gogen-avro/schema/canonical"
 )
 
-{{ if ne .Doc "" }}// {{ .Doc}}{{ end }}  
+var {{ .Name }}UID []byte
```
Comment (actgardner) [inline generator/flat/templates/record.go]: This should be a `const` we compute when we generate the code, rather than re-computing it at package initialization time.

File: schema/canonical/crc.go [hunk 2]
```diff
@@ -46,12 +27,53 @@ func init() {
...
 		fp = (fp >> 8) ^ FP_TABLE[(fp^uint64(d))&0xff]
 	}
-	output := bytes.NewBuffer(make([]byte, 0, 8))
-	binary.Write(output, binary.LittleEndian, fp)
-	return hex.EncodeToString(output.Bytes())
+	err := binary.Write(writer, binary.LittleEndian, fp)
+	return err
+}
+
+func AvroCalcSchemaUID(schema string) []byte {
```
Comment (actgardner) [inline schema/canonical/crc.go]: If we generate the UID when we're generating code we also don't need this method to re-parse the schema.

File: schema/canonical/crc_test.go [hunk 3]
```diff
@@ -1,24 +1,52 @@
...
...a.string\":\"String\"},\"default\":\"\"},{\"name\":\"validityPeriod\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"default\":\"\"},{\"name\":\"registeredDelivery\",\"type\":{\"type\":\"int\",\"java-class\":\"java.lang.Byte\"},\"default\":0},{\"name\":\"replaceIfPresent\",\"type\":{\"type\":\"int\",\"java-class\":\"java.lang.Byte\"},\"default\":0},{\"name\":\"dataCoding\",\"type\":{\"type\":\"int\",\"java-class\":\"java.lang.Byte\"},\"default\":0},{\"name\":\"defaultMsgId\",\"type\":{\"type\":\"int\",\"java-class\":\"java.lang.Byte\"},\"default\":0},{\"name\":\"shortMessage\",\"type\":{\"type\":\"bytes\",\"java-class\":\"[B\"},\"default\":\"ÿ\"},{\"name\":\"optionalParameters\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Tlv\",\"fields\":[{\"name\":\"tag\",\"type\":{\"type\":\"int\",\"java-class\":\"java.lang.Short\"},\"default\":0},{\"name\":\"value\",\"type\":{\"type\":\"bytes\",\"java-class\":\"[B\"},\"default\":\"ÿ\"},{\"name\":\"tagName\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"default\":\"\"}]},\"java-class\":\"java.util.ArrayList\"},\"default\":[]}]},\"default\":{}}],\"default\":{}}", "0c947f601de7ce84"},
```
Comment (actgardner) [inline schema/canonical/crc_test.go]: This test is just for the CRC64 computation, I don't think it makes sense to add testing the canonical form of a schema here.

File: generator/flat/templates/record.go (hunk 1 above)
Comment (dhiaayachi) [reply generator/flat/templates/record.go]: I will have to code generate the byte array as a const. I'm not very familiar with how the code generation work in go and the doc resources for it are very limited. I will try to figure it out otherwise I will ask for some help from you.

File: schema/canonical/crc_test.go (hunk 3 above)
Comment (dhiaayachi) [reply schema/canonical/crc_test.go]: I Agree, I will separate the canonical from the crc and hardcode the canonical form in the crc test.

File: schema/canonical/crc.go (hunk 2 above)
Comment (dhiaayachi) [reply schema/canonical/crc.go]: 👍

File: generator/flat/templates/record.go (hunk 1 above)
Comment (actgardner) [reply generator/flat/templates/record.go]: The simplest thing to do would be to add a method to the record definition type, and then call it in the template like `{{ .AvroUID }}`

File: schema/canonical/crc.go [hunk 4]
```diff
@@ -46,12 +27,53 @@ func init() {
...
+}
+
+func AvroConsumeHeader(r io.Reader) error {
+	buf := make([]byte, 10)
+	n, err := r.Read(buf)
+	if n < 10 || err != nil {
+		return errors.New("message length <2")
+	}
+	return nil
 }
```
Comment (actgardner) [inline schema/canonical/crc.go]: This file is all about CRC computation, I think it would make more sense to put the single-object encoding magic number and framing code in the soe package.

File: singleobject/reader.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
+package singleobject
```
Comment (actgardner) [inline singleobject/reader.go]: I think the convention for naming would be `single_object`, but I'd also be ok with `soe` for this package.

File: singleobject/reader.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	"bytes"
+	"errors"
+)
+
+type soReader struct {
+	r      *bytes.Buffer
+	header []byte
+}
+
+func NewReader(buf *bytes.Buffer) *soReader {
```
Comment (actgardner) [inline singleobject/reader.go]: I know I asked for a reader/writer interface but I realize it probably makes more sense for this method to have a signature like:

```
func ReadHeader(r io.Reader) (fingerprint string, err error)
```

And then users would be able to parse the header and switch on it to figure out what schema it represents like:

```
fingerprint, err := ReadHeader(r)
switch fingerprint {
  case SchemaOneFingerprint:
    DeserializeSchemaOne(r)
  case SchemaTwoFingerprint:
    DeserializeSchemaTwo(r)
}
```

File: singleobject/writer.go [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+	"bytes"
+	"github.com/actgardner/gogen-avro/schema/canonical"
+)
+
+type soWriter struct {
+	w      *bytes.Buffer
+	header []byte
+}
+
+func NewWriter(buf *bytes.Buffer, header []byte) *soWriter {
```
Comment (actgardner) [inline singleobject/writer.go]: Likewise this probably doesn't have to be a writer but just a convenience method that accepts an interface and encodes it to an io.Writer:

```

type Encodable interface {
	Serialize(io.Writer) error
	AvroUID() string
}

func Write(e Encodable, w io.Writer) (error) 
``` 

See the container package for a similar idea:
https://github.com/actgardner/gogen-avro/blob/master/container/writer.go#L94
https://github.com/actgardner/gogen-avro/blob/master/container/avro_record.go#L10-L13

File: generator/flat/templates/record.go (hunk 1 above)
Comment (dhiaayachi) [reply generator/flat/templates/record.go]: I was working on it and did exactly that. The issue I'm facing is that I need the parser to calculate the canonical form and generate the UID. it create a cyclic import issue schema-->parser-->schema. Any idea on how to resolve that?

Comment (actgardner) [reply generator/flat/templates/record.go]: I pushed a commit to master that adds the hex-encoded CRC64 for the canonical schema as a constant to generated record types, it took a little bit more finagling than expected.

Comment (dhiaayachi) [reply generator/flat/templates/record.go]: I see, is it ok if I change the type to a slice of byte? If it's a hex string extra conversion is needed for each read and write.

Comment (dhiaayachi) [reply generator/flat/templates/record.go]: I merged your commit and made the changes to generate the fingerprint in []byte

File: schema/canonical/crc.go (hunk 2 above)
Comment (dhiaayachi) [reply schema/canonical/crc.go]: func removed

File: singleobject/reader.go (hunk 5 above)
Comment (dhiaayachi) [reply singleobject/reader.go]: renaming done

File: schema/canonical/crc.go (hunk 4 above)
Comment (dhiaayachi) [reply schema/canonical/crc.go]: done

File: singleobject/reader.go (hunk 5 above)
Comment (dhiaayachi) [reply singleobject/reader.go]: done

File: schema/canonical/crc_test.go (hunk 3 above)
Comment (dhiaayachi) [reply schema/canonical/crc_test.go]: reverted the tests

File: generator/flat/templates/lookup.go [hunk 8]
```diff
@@ -57,3 +58,15 @@ func Evaluate(templateStr string, obj interface{}) (string, error) {
 
 	return buf.String(), nil
 }
+
+func convertByteToInitForm(b []byte) string {
```
Comment (actgardner) [inline generator/flat/templates/lookup.go]: A Go string can encode any byte slice just by using the `%q` formatting verb, like `fmt.Sprintf("%q", []byte{0x01, 0xfd})`. Ex. https://play.golang.org/p/V-JgdJce9Eu

File: generator/flat/templates/record.go [hunk 9]
```diff
@@ -22,7 +22,7 @@ type {{ .Name }} struct {
 {{ end }}
 }
 
-const {{ .Name }}AvroCRC64Fingerprint = "{{ definitionFingerprint . }}"
+var {{ .Name }}AvroCRC64Fingerprint = {{ definitionFingerprint . }}
```
Comment (actgardner) [inline generator/flat/templates/record.go]: If we can't make this a const because it's a byte slice, let's just put the value inside the method definition? Right now a user could reassign the fingerprint variable for a generated struct which would lead to weird behaviour.

File: schema/canonical/crc.go [hunk 10]
```diff
@@ -3,38 +3,14 @@ package canonical
...
-    FP_TABLE[i] = fp;
-  }
-}
-*/
-
 var FP_TABLE []uint64
 
 const EMPTY uint64 = 0xc15d213aa4d7a795
 
+var HeaderV1 = []byte{0xC3, 0x01}
```
Comment (actgardner) [inline schema/canonical/crc.go]: This should be in the soe package since it's part of the framing and not part of the CRC calculation.

File: schema/canonical/crc_test.go [hunk 11]
```diff
@@ -1,24 +1,23 @@
...
 	"github.com/stretchr/testify/assert"
+	"testing"
 )
 
 func TestAvroCRC64Fingerprint(t *testing.T) {
 	cases := []struct {
 		schema      string
 		fingerprint string
 	}{
-		{`"int"`, "8f5c393f1ad57572"},
```
Comment (actgardner) [inline schema/canonical/crc_test.go]: Why are these removed?

File: schema/canonical/crc_test.go [hunk 12]
```diff
@@ -1,24 +1,23 @@
 package canonical
 
 import (
-	"testing"
```
Comment (actgardner) [inline schema/canonical/crc_test.go]: This is a deliberate convention - stdlib imports are in a separate block from 3rd party imports. `testing` and `encoding/hex` are stdlib, but testify is a 3rd party library.

File: test/alias-record/schema_test.go [hunk 13]
```diff
@@ -13,22 +14,22 @@ const fixtureJson = `
...
-      "BoolField": true, 
+      "StringField": "abcdghejw",
+      "BoolField": true,
       "BytesField": "VGhpcyBpcyBhIHRlc3Qgc3RyaW5n"
     }
   },
   {
     "OtherField": {
-      "StringField": "jdnwjkendwedddedee", 
-      "BoolField": true,
```
Comment (actgardner) [inline test/alias-record/schema_test.go]: Removing trailing whitespace is helpful, but it adds a lot of noise to this PR. For things like renaming variables and removing whitespace it would be easier for me to review a separate PR that doesn't also add new features.

**Comment by actgardner** [comment]
I don't think we need to use the SOE package for the integration tests. It should be sufficient to write a couple unit tests in the soe package for serializing and deserializing headers.

File: generator/flat/templates/lookup.go (hunk 8 above)
Comment (dhiaayachi) [reply generator/flat/templates/lookup.go]: I did the change to use this. As a bonus we can use the global as Const, because I can define it as string and do casting to []byte at the function level.

File: generator/flat/templates/record.go (hunk 9 above)
Comment (dhiaayachi) [reply generator/flat/templates/record.go]: See my comment about the use of %q

File: schema/canonical/crc_test.go (hunk 12 above)
Comment (dhiaayachi) [reply schema/canonical/crc_test.go]: I use goland as IDE and it have tendency to do things behind my back

Comment (dhiaayachi) [reply schema/canonical/crc_test.go]: I will revert

File: test/alias-record/schema_test.go (hunk 13 above)
Comment (dhiaayachi) [reply test/alias-record/schema_test.go]: > I don't think we need to use the SOE package for the integration tests. It should be sufficient to write a couple unit tests in the soe package for serializing and deserializing headers.

I would advice to add soe to at least one test to cover soe integration with the serialization/deserialization though.

File: singleobject/reader.go (hunk 6 above)
Comment (dhiaayachi) [reply singleobject/reader.go]: did the change code is much simpler

File: singleobject/writer.go (hunk 7 above)
Comment (dhiaayachi) [reply singleobject/writer.go]: did the change code is much simpler

**Comment by dhiaayachi** [comment]
I did all of the change required in the review, please carry on with the review when you have time and let me know if any other change is needed.
I did not do the integration test for soe, waiting for your comment on that.

**Comment by actgardner** [comment]
Looks good to me, thanks for your work on this! I'll merge it to master and add some docs and example code shortly.

**Comment by dhiaayachi** [comment]
Thank you for all of your comments and code review. Nice to work with you!!
Do you have an ETA for a new release, I would like to start using it in my project asap.

**Comment by actgardner** [comment]
If you're using go modules you can manually specify any git commit by SHA like `go get github.com/actgardner/gogen-avro@<SHA>`. So you don't need to wait for a release to use this. I'll cut one once the docs are ready.

**Comment by dhiaayachi** [comment]
ok thank you, I will try that.
